File: 2116-check-if-a-parentheses-string-can-be-valid/2116-check-if-a-parentheses-string-can-be-valid.py
class Solution:
  def canBeValid(self, s, locked):
    index, stack = [], []
    for i in range(len(s)):
      ch = s[i]
      if locked[i] == '0':
        stack.append('')
      elif ch == ')':
        if not len(stack):
          return False
        if len(index):
          stack.pop(index.pop())
        else:
          stack.pop()
      else:
        index.append(len(stack))
        stack.append('(')
    #   print(ch, stack)
    # print(stack)
    opened, balanced = 0, True
    for ch in stack:
      if ch == '(':
        opened += 1
      elif ch == '':
        if opened:
          opened -= 1
        else:
          balanced = not balanced
    # print(opened, balanced)
    return opened == 0 and balanced

File: 2116-check-if-a-parentheses-string-can-be-valid/2116-check-if-a-parentheses-string-can-be-valid.ts
function canBeValid(s: string, locked: string): boolean {
  const index: number[] = [],
    stack: boolean[] = [];
  for (let i = 0; i < s.length; i++) {
    if (locked[i] === '0') {
      stack.push(false);
    } else if (s[i] === ')') {
      if (!stack.length) return false;
      if (index.length) {
        stack[index.pop()] = false;
      }
      stack.pop();
    } else {
      index.push(stack.length);
      stack.push(true);
    }
  }
  let opened = 0,
    balanced = true;
  for (const item of stack) {
    if (item) {
      opened++;
    } else if (opened) {
      opened--;
    } else {
      balanced = !balanced;
    }
  }
  return opened == 0 && balanced;
};